selfdrive/boardd/tests/test_boardd_usbprotocol.cc

```diff
@@ -0,0 +1,127 @@
+#define CATCH_CONFIG_MAIN
+#define CATCH_CONFIG_ENABLE_BENCHMARKING
+#include <random>
+
+#include "catch2/catch.hpp"
+#include "cereal/messaging/messaging.h"
+#include "selfdrive/boardd/panda.h"
+
+const int CHUNK_SIZE = 64;
+const unsigned char dlc_to_len[] = {0U, 1U, 2U, 3U, 4U, 5U, 6U, 7U, 8U, 12U, 16U, 20U, 24U, 32U, 48U, 64U};
+
+int random_int(int min, int max) {
+  std::random_device dev;
+  std::mt19937 rng(dev());
+  std::uniform_int_distribution<std::mt19937::result_type> dist(min, max);
+  return dist(rng);
+}
+
+struct PandaTest : public Panda {
+  PandaTest(uint32_t bus_offset, int can_list_size);
+  void test_can_send();
+  void test_can_recv();
+
+  std::map<int, std::string> test_data;
+  int can_list_size = 0;
+  int total_pakets_size = 0;
+  MessageBuilder msg;
+  capnp::List<cereal::CanData>::Reader can_data_list;
+};
+
+PandaTest::PandaTest(uint32_t bus_offset_, int can_list_size) : can_list_size(can_list_size), Panda(bus_offset_) {
+  // prepare test data
+  for (int i = 0; i < std::size(dlc_to_len); ++i) {
+    std::random_device rd;
+    std::independent_bits_engine<std::default_random_engine, CHAR_BIT, unsigned char> rbe(rd());
+
+    int data_len = dlc_to_len[i];
+    std::string bytes(data_len, '\0');
+    std::generate(bytes.begin(), bytes.end(), std::ref(rbe));
+    test_data[data_len] = bytes;
+  }
+
+  // generate can messages for this panda
+  int other_panda_message_cnt = 3;
+  auto can_list = msg.initEvent().initSendcan(can_list_size + other_panda_message_cnt);
+  for (uint8_t i = 0; i < can_list_size; ++i) {
+    auto can = can_list[i];
+    uint32_t id = random_int(0, std::size(dlc_to_len) - 1);
+    const std::string &dat = test_data[dlc_to_len[id]];
+    can.setAddress(i);
+    can.setSrc(bus_offset);
```

Comment:
```suggestion
    can.setSrc(random_int(0, 3) + bus_offset);
```

And then loop for making messages can be deleted. So we test bus_offset = 0 first and then 4(for extra panda)